File: day3/solution3.2.ts
import * as fs from "fs";

function calculateEnabledMultiplications(filePath: string): number {
    const memory = fs.readFileSync(filePath, "utf-8");

    // Regex patterns
    const mulPattern = /mul\((\d+),(\d+)\)/g;
    const togglePattern = /do\(\)|don't\(\)/g;

    // Extract instructions
    const instructions = [
        ...Array.from(memory.matchAll(togglePattern)).map(match => ({
            index: match.index!,
            type: match[0]
        })),
        ...Array.from(memory.matchAll(mulPattern)).map(match => ({
            index: match.index!,
            type: "mul",
            x: parseInt(match[1], 10),
            y: parseInt(match[2], 10)
        }))
    ].sort((a, b) => a.index - b.index);

    // Initialize state
    let mulEnabled = true;
    let total = 0;

    for (const instr of instructions) {
        if (instr.type === "do()") {
            mulEnabled = true;
        } else if (instr.type === "don't()") {
            mulEnabled = false;
        } else if (instr.type === "mul" && mulEnabled) {
            total += instr.x * instr.y;
        }
    }

    return total;
}

// Example usage
const result = calculateEnabledMultiplications("input.txt");
console.log(`Total sum of enabled multiplications: ${result}`);

File: day3/solution3.2.py
import re

def calculate_enabled_multiplications(file_path):
    # Read the input file
    with open(file_path, 'r') as f:
        memory = f.read()

    # Regex to find valid mul(X,Y) instructions
    mul_pattern = re.compile(r"mul\((\d+),(\d+)\)")
    # Regex to find do() and don't() instructions
    toggle_pattern = re.compile(r"(do\(\)|don't\(\))")

    # Find all instructions in the order they appear
    instructions = toggle_pattern.findall(memory) + mul_pattern.findall(memory)
    instructions = sorted(
        [(m.start(), m.group()) for m in toggle_pattern.finditer(memory)] +
        [(m.start(), ("mul", int(m.group(1)), int(m.group(2)))) for m in mul_pattern.finditer(memory)],
        key=lambda x: x[0]
    )

    # Initialize state
    mul_enabled = True
    total = 0

    for _, instr in instructions:
        if instr == "do()":
            mul_enabled = True
        elif instr == "don't()":
            mul_enabled = False
        elif isinstance(instr, tuple) and instr[0] == "mul":
            if mul_enabled:
                total += instr[1] * instr[2]

    return total

# Example usage
result = calculate_enabled_multiplications("input.txt")
print(f"Total sum of enabled multiplications: {result}")